### PR TITLE
Fix agent MSI and service tests in Juniper CI

### PR DESCRIPTION
**Pull request by m-kostrzewa**
This is based off of @krdln's PR.

Changes:
* We agreed with @jablonskim to install required artifacts explicitely for each test. In "old CI", artifacts were already installed. So I added installation/uninstallation of required artifacts in `BeforeAll/AfterAll` blocks.
* In `vRouterAgentMSIInstaller.Tests.ps1` there was a missing test scenario; Agent should *crash* when started and Extension is not running. So the fact that we don't check for the negative scenario in MSI tests could potentially lead to a regression.
* Moved MSI related helpers to separate file.
* Enabled tests that don't require docker driver in CI.

### CODE REVIEW

**Comment by m-kostrzewa** [comment]
The branch conflict will be resolved after @krdln's is merged.

**Comment by krdln** [comment]
Can you rebase instead of merging, so the commit history shows only relevant commits?

File: CIScripts/Test/Tests/VTestScenarios.Tests.ps1 [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
 )
 
 . $PSScriptRoot\..\Utils\CommonTestCode.ps1
+. $PSScriptRoot\..\Utils\ComponentsInstallation.ps1
 . $PSScriptRoot\..\TestConfigurationUtils.ps1
 . $PSScriptRoot\..\..\Common\VMUtils.ps1
 . $PSScriptRoot\..\PesterHelpers\PesterHelpers.ps1
 
-. $PSScriptRoot\..\GetTestConfigurationJuni.ps1
+. $PSScriptRoot\..\($TestConfigurationFile)
```
Comment (krdln) [inline CIScripts/Test/Tests/VTestScenarios.Tests.ps1]: $TestConfigurationFile is passed as a full path, so just `. $TestConfigurationFile` should work.

File: CIScripts/Test/Utils/ComponentsInstallation.ps1 [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    Write-Host "Installing vRouter utility tools"
+    Invoke-MsiExec -Session $Session -Path "C:\Artifacts\utils.msi"
+}
+
+function Uninstall-Utils {
+    Param ([Parameter(Mandatory = $true)] [PSSessionT] $Session)
+
+    Write-Host "Uninstalling vRouter utility tools"
+    Invoke-MsiExec -Uninstall -Session $Session -Path "C:\Artifacts\utils.msi"
+}
```
Comment (krdln) [inline CIScripts/Test/Utils/ComponentsInstallation.ps1:64]: Did you mean to copy these functions from `CIScripts/Test/Utils/CommonTestCode.ps1` or move them?

File: CIScripts/Test/Tests/VTestScenarios.Tests.ps1 [hunk 3]
```diff
@@ -1,13 +1,15 @@
 Param (
-    [Parameter(Mandatory=$true)] [string] $TestbedAddr
+    [Parameter(Mandatory=$true)] [string] $TestbedAddr,
+    [Parameter(Mandatory=$true)] [string] $TestConfigurationFile
```
Comment (krdln) [inline CIScripts/Test/Tests/VTestScenarios.Tests.ps1]: Nit: trailing whitespace

**Comment by krdln** [comment]
Retried with `COMPONENTS_TO_BUILD = "None"`, [but it fails](http://10.84.12.26:8080/job/WinContrail/job/winci-server2016-contrail-windows-ci/488/console).

File: .gitignore [hunk 4]
```diff
@@ -15,3 +15,6 @@ CIScripts/JenkinsTesterDSC/JenkinsTester
 
 ### Python ###
 *.pyc
+
+### Visual Studio Code
+.vscode/
```
Comment (krdln) [inline .gitignore]: Why was this added?

File: CIScripts/Test/Tests/vRouterAgentMSIInstaller.Tests.ps1 [hunk 5]
```diff
@@ -1,37 +1,55 @@
...
+
+    function Test-AgentMSIBehaviourCorrect {
+        Install-Agent -Session $Session
+        Eventually {
+            Get-AgentServiceStatus -Session $Session | Should Be "Stopped"
+        } -Duration 3
+
+        Uninstall-Agent -Session $Session
+        Eventually {
+            Get-AgentServiceStatus -Session $Session | Should Throw
```
Comment (krdln) [inline CIScripts/Test/Tests/vRouterAgentMSIInstaller.Tests.ps1]: 1. `Should Throw` always needs `{}`
2. You've refactored `Get-AgentServiceStatus` never to throw so you wanted `Should Be Null` probably.

File: CIScripts/Test/Tests/vRouterAgentMSIInstaller.Tests.ps1 [hunk 6]
```diff
@@ -1,37 +1,55 @@
...
+    Context "when vRouter Forwarding Extension is running" {
+        It "registers/unregisters Agent service and never enables" {
+            Test-AgentMSIBehaviourCorrect
+        }
+
+        BeforeEach {
+            Install-Extension -Session $Session
+            Enable-VRouterExtension -Session $Session -AdapterName $TestConf.AdapterName `
+                -VMSwitchName $TestConf.VMSwitchName `
+                -ForwardingExtensionName $TestConf.ForwardingExtensionName
```
Comment (krdln) [inline CIScripts/Test/Tests/vRouterAgentMSIInstaller.Tests.ps1:50]: Do we need those parameters to `Enable-VRouterExtension`? We always call it with the same set of params.

File: CIScripts/Test/Tests/vRouterAgentService.Tests.ps1 [hunk 7]
```diff
@@ -91,7 +93,6 @@ Describe "vRouter Agent service" {
 
     BeforeEach {
         Initialize-DriverAndExtension -Session $Session -TestConfiguration $TestConf
```
Comment (krdln) [inline CIScripts/Test/Tests/vRouterAgentService.Tests.ps1:99]: Are we sure we need docker driver for these tests?

**Review by krdln — CHANGES_REQUESTED** [review]
MK + KR

**Comment by krdln** [comment]
`After-Each` runs after every `It` in every context, not just after every context, right?

File: CIScripts/Test/Utils/ComponentsInstallation.ps1 (hunk 2 above)
Comment (krdln) [reply CIScripts/Test/Utils/ComponentsInstallation.ps1:64]: Also, why move them out from `CommonTestCode`? Is your plan to split this file?

File: .gitignore (hunk 4 above)
Comment (m-kostrzewa) [reply .gitignore]: Why not? :D Most of the team uses vscode

**Comment by krdln** [comment]
[Log with artifacts from `C:\WindowsCI-Artifacts`](http://10.84.12.26:8080/job/WinContrail/job/winci-server2016-contrail-windows-ci/514/console)

Let's wait for https://github.com/codilime/contrail-windows-ci/pull/112 and enabling `COPY_DISABLED_ARTIFACTS`. Also, there's another problem with partial build – `Deploy` script currently [only copies artifacts for components present in `COMPONENTS_TO_BUILD`](https://github.com/codilime/contrail-windows-ci/blob/305c852cb7f4cc3643ddb6a8567825ad26333366/CIScripts/Deploy/Deployment.ps1#L39). Wee need to change it somehow.

**Comment by krdln** [comment]
requires https://github.com/codilime/contrail-windows-ci/pull/120

**Comment by m-kostrzewa** [comment]
recheck no bug

File: CIScripts/Test/Tests/vRouterAgentMSIInstaller.Tests.ps1 (hunk 6 above)
Comment (m-kostrzewa) [reply CIScripts/Test/Tests/vRouterAgentMSIInstaller.Tests.ps1:50]: Eventual fix is to have a `Testbed` class, that would store stuff like name of VM switch as a field. But not in scope of this PR IMO.

File: CIScripts/Test/PesterHelpers/PesterHelpers.Tests.ps1 [hunk 8]
```diff
@@ -10,15 +10,6 @@ Describe "PesterHelpers" {
             { Consistently { $true | Should Not Be $true } -Duration 3 } | Should Throw
         }
 
-        It "works with inner exceptions" {
```
Comment (m-kostrzewa) [inline CIScripts/Test/PesterHelpers/PesterHelpers.Tests.ps1:13]: Removed these tests because the behaviour was inconsistent.

File: CIScripts/Test/TestConfigurationUtils.ps1 [hunk 9]
```diff
@@ -218,7 +218,11 @@ function Get-AgentServiceStatus {
     $Status = Invoke-Command -Session $Session -ScriptBlock {
         (Get-Service "ContrailAgent").Status
     }
-    return $Status.Value
```
Comment (m-kostrzewa) [inline CIScripts/Test/TestConfigurationUtils.ps1]: StrictMode helped us here :)

File: CIScripts/Test/Utils/ComponentsInstallation.ps1 (hunk 2 above)
Comment (m-kostrzewa) [reply CIScripts/Test/Utils/ComponentsInstallation.ps1:64]: Oversight on my part, I wanted to move those functions to separate file.

Comment (m-kostrzewa) [reply CIScripts/Test/Utils/ComponentsInstallation.ps1:64]: > Also, why move them out from CommonTestCode? Is your plan to split this file?

yes. They form a thematic subgroup which is not related to any other functions in this file. Common*, Utils*, Misc* is very often an antipattern.

Comment (m-kostrzewa) [reply CIScripts/Test/Utils/ComponentsInstallation.ps1:64]: E.g. if I want to install stuff using MSI, I have to import the whole file that has functions that are useless to me, like pinging containers.

File: CIScripts/Test/Tests/vRouterAgentService.Tests.ps1 [hunk 10]
```diff
@@ -96,7 +96,10 @@ Describe "vRouter Agent service" {
     }
 
     BeforeEach {
-        Initialize-DriverAndExtension -Session $Session -TestConfiguration $TestConf
+        Install-Extension -Session $Session
```
Comment (krdln) [inline CIScripts/Test/Tests/vRouterAgentService.Tests.ps1]: It's already installed in `BeforeAll`, not sure if it's needed here.

File: CIScripts/Test/TestRunner.ps1 [hunk 11]
```diff
@@ -43,10 +43,7 @@ function Invoke-TestScenarios {
 
     $TestsBlacklist = @(
         # Requires Docker Driver
-        "vRouterAgentService.Tests.ps1",
-
-        # Requires Agent
-        "vRouterAgentMSIInstaller.Tests.ps1"
+        "vRouterAgentService.Tests.ps1"
```
Comment (krdln) [inline CIScripts/Test/TestRunner.ps1]: It could be removed from blacklist, as you've changed it not to require docker driver.

File: CIScripts/Test/TestConfigurationUtils.ps1 [hunk 12]
```diff
@@ -216,9 +216,13 @@ function Disable-AgentService {
 function Get-AgentServiceStatus {
     Param ([Parameter(Mandatory = $true)] [PSSessionT] $Session)
     $Status = Invoke-Command -Session $Session -ScriptBlock {
-        (Get-Service "ContrailAgent").Status
+        (Get-Service "ContrailAgent" -ErrorAction SilentlyContinue).Status
+    }
+    if($Status) {
```
Comment (krdln) [inline CIScripts/Test/TestConfigurationUtils.ps1]: Please add a space after `if`

**Review by krdln — CHANGES_REQUESTED** [review]

File: CIScripts/Test/Tests/vRouterAgentMSIInstaller.Tests.ps1 (hunk 6 above)
Comment (krdln) [reply CIScripts/Test/Tests/vRouterAgentMSIInstaller.Tests.ps1:50]: Well, we kind of have this class already, and it's called `TestConfiguration`. But it's fine not to change it in this PR.

**Review by katrybacka — DISMISSED** [review]

**Review by krdln — DISMISSED** [review]

File: CIScripts/Test/Utils/ComponentsInstallation.ps1 [hunk 13]
```diff
@@ -62,3 +62,24 @@ function Uninstall-Utils {
     Write-Host "Uninstalling vRouter utility tools"
     Invoke-MsiExec -Uninstall -Session $Session -Path "C:\Artifacts\utils.msi"
 }
+
+function Install-Utils {
+    Param ([Parameter(Mandatory = $true)] [PSSessionT] $Session)
+
+    Write-Host "Installing vRouter utility tools"
+    Invoke-MsiExec -Session $Session -Path "C:\Artifacts\utils.msi"
+}
```
Comment (krdln) [inline CIScripts/Test/Utils/ComponentsInstallation.ps1]: You've created two `Install-Utils` :P

**Review by krdln — CHANGES_REQUESTED** [review]

**Comment by krdln** [comment]
Docker driver requires working controller, so we have to blacklist the test for now and enable it after https://github.com/codilime/contrail-windows-ci/pull/121.

**Review by krdln — APPROVED** [review]
